File: src/main/java/com/ocelot/vehicle/jei/plugin/workstation/VehicleRecipeWrapper.java
//package com.ocelot.vehicle.jei.plugin.workstation;
//
//import com.mrcrayfish.vehicle.common.entity.PartPosition;
//import com.mrcrayfish.vehicle.entity.EngineType;
//import com.mrcrayfish.vehicle.entity.EntityPoweredVehicle;
//import com.mrcrayfish.vehicle.entity.EntityVehicle;
//import com.mrcrayfish.vehicle.entity.VehicleProperties;
//import com.ocelot.vehicle.jei.VehicleModJei;
//import mezz.jei.api.IGuiHelper;
//import mezz.jei.api.gui.IDrawableStatic;
//import mezz.jei.api.ingredients.IIngredients;
//import mezz.jei.api.ingredients.VanillaTypes;
//import mezz.jei.api.recipe.IRecipeWrapper;
//import net.minecraft.client.Minecraft;
//import net.minecraft.client.renderer.GlStateManager;
//import net.minecraft.client.resources.I18n;
//import net.minecraft.item.ItemStack;
//import net.minecraft.util.text.TextFormatting;
//
//import javax.annotation.Nonnull;
//import java.util.ArrayList;
//import java.util.List;
//import java.util.function.Function;
//
//public class VehicleRecipeWrapper implements IRecipeWrapper
//{
//
//    public static final float SCALE = 20;
//    public static final int MATERIAL_SLOTS = 8;
//
//    private final IDrawableStatic slot;
//    private final IDrawableStatic slotIconColor;
//    private final IDrawableStatic slotIconNone;
//
//    private final List<List<ItemStack>> itemInputs;
//    private final VehicleIngredient vehicleOutput;
//
//    VehicleRecipeWrapper(IGuiHelper guiHelper, List<List<ItemStack>> itemInputs, VehicleIngredient vehicleOutput)
//    {
//        this.slot = guiHelper.createDrawable(VehicleModJei.RECIPE_GUI_VEHICLE, 176, 43, 18, 18);
//        this.slotIconColor = guiHelper.createDrawable(VehicleModJei.RECIPE_GUI_VEHICLE, 194, 43, 16, 16);
//        this.slotIconNone = guiHelper.createDrawable(VehicleModJei.RECIPE_GUI_VEHICLE, 210, 43, 16, 16);
//
//        this.itemInputs = itemInputs;
//        this.vehicleOutput = vehicleOutput;
//    }
//
//    @Override
//    public void getIngredients(IIngredients ingredients)
//    {
//        ingredients.setInputLists(VanillaTypes.ITEM, this.itemInputs);
//        ingredients.setOutput(VanillaTypes.ITEM, this.vehicleOutput.crate);
//    }
//
//    @Override
//    public void drawInfo(Minecraft minecraft, int recipeWidth, int recipeHeight, int mouseX, int mouseY)
//    {
//        EntityVehicle vehicle = vehicleOutput.vehicle;
//
//        if (SlotType.COLOR.isApplicable(vehicle))
//        {
//            slotIconColor.draw(minecraft, 1, 65);
//        }
//        else
//        {
//            slotIconNone.draw(minecraft, 1, 65);
//        }
//        if (!SlotType.ENGINE.isApplicable(vehicle))
//        {
//            slotIconNone.draw(minecraft, 21, 65);
//        }
//        if (!SlotType.WHEELS.isApplicable(vehicle))
//        {
//            slotIconNone.draw(minecraft, 41, 65);
//        }
//
//        for (int i = 0; i < MATERIAL_SLOTS; i++)
//        {
//            slot.draw(minecraft, i * 18, 95);
//        }
//
//        minecraft.fontRenderer.drawString(I18n.format(VehicleModJei.MOD_ID + ".category.vehicle.materials"), 0, 85, 4210752);
//
//        GlStateManager.enableDepth();
//        GlStateManager.color(1, 1, 1, 1);
//        GlStateManager.translate(72, 50, 100);
//        GlStateManager.rotate(-5, 1, 0, 0);
//        GlStateManager.rotate(Minecraft.getMinecraft().player.ticksExisted + minecraft.getRenderPartialTicks(), 0f, 1f, 0f);
//        GlStateManager.scale(SCALE, -SCALE, SCALE);
//
//        VehicleProperties properties = VehicleProperties.getProperties(vehicleOutput.vehicle.getClass());
//        if (properties != null)
//        {
//            PartPosition position = properties.getDisplayPosition();
//            //Apply vehicle rotations, translations, and scale
//            GlStateManager.scale(position.getScale(), position.getScale(), position.getScale());
//            GlStateManager.rotate((float) position.getRotX(), 1, 0, 0);
//            GlStateManager.rotate((float) position.getRotY(), 0, 1, 0);
//            GlStateManager.rotate((float) position.getRotZ(), 0, 0, 1);
//            GlStateManager.translate(position.getX(), position.getY(), position.getZ());
//        }
//        minecraft.getRenderManager().getEntityClassRenderObject(vehicleOutput.vehicle.getClass()).doRender(vehicle, 0, 0, 0, 0, 0);
//        GlStateManager.disableDepth();
//    }
//
//    @Nonnull
//    @Override
//    public List<String> getTooltipStrings(int mouseX, int mouseY)
//    {
//        List<String> tooltip = new ArrayList<>();
//
//        addSlotTooltip(tooltip, 0, 64, mouseX, mouseY, SlotType.COLOR);
//        addSlotTooltip(tooltip, 20, 64, mouseX, mouseY, SlotType.ENGINE);
//        addSlotTooltip(tooltip, 40, 64, mouseX, mouseY, SlotType.WHEELS);
//
//        return tooltip;
//    }
//
//    private void addSlotTooltip(List<String> tooltip, int slotX, int slotY, int mouseX, int mouseY, SlotType type)
//    {
//        if (mouseX >= slotX && mouseX < slotX + 18 && mouseY >= slotY && mouseY < slotY + 18)
//        {
//            if (!type.isApplicable(vehicleOutput.vehicle))
//            {
//                tooltip.add(TextFormatting.WHITE + I18n.format("vehicle.tooltip." + type.registryName));
//                tooltip.add(TextFormatting.GRAY + I18n.format("vehicle.tooltip.not_applicable"));
//            }
//            else if (type == SlotType.COLOR)
//            {
//                tooltip.add(TextFormatting.AQUA + I18n.format("vehicle.tooltip.optional"));
//                tooltip.add(TextFormatting.GRAY + I18n.format("vehicle.tooltip." + type.registryName));
//            }
//        }
//    }
//
//    private enum SlotType
//    {
//        COLOR("paint_color", EntityVehicle::canBeColored),
//        ENGINE("engine", vehicle -> vehicle instanceof EntityPoweredVehicle && ((EntityPoweredVehicle) vehicle).getEngineType() != EngineType.NONE),
//        WHEELS("wheels", vehicle -> vehicle instanceof EntityPoweredVehicle && ((EntityPoweredVehicle) vehicle).canChangeWheels());
//
//        private final String registryName;
//        private final Function<EntityVehicle, Boolean> applicable;
//
//        SlotType(String registryName, Function<EntityVehicle, Boolean> applicable)
//        {
//            this.registryName = registryName;
//            this.applicable = applicable;
//        }
//
//        public boolean isApplicable(EntityVehicle vehicle)
//        {
//            return applicable.apply(vehicle);
//        }
//
//        public String getRegistryName()
//        {
//            return registryName;
//        }
//    }
//}

File: src/main/java/com/ocelot/vehicle/jei/plugin/workstation/VehicleRecipeWrapper.kt
package com.ocelot.vehicle.jei.plugin.workstation

import com.mrcrayfish.vehicle.client.gui.GuiWorkstation
import com.mrcrayfish.vehicle.entity.EngineType
import com.mrcrayfish.vehicle.entity.EntityPoweredVehicle
import com.mrcrayfish.vehicle.entity.EntityVehicle
import com.ocelot.vehicle.jei.VehicleModJei
import mezz.jei.api.IGuiHelper
import mezz.jei.api.ingredients.IIngredients
import mezz.jei.api.ingredients.VanillaTypes
import mezz.jei.api.recipe.IRecipeWrapper
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.resources.I18n
import net.minecraft.item.ItemStack
import net.minecraft.util.text.TextFormatting
import java.util.function.Function

class VehicleRecipeWrapper(guiHelper: IGuiHelper, private val itemInputs: List<List<ItemStack>>, private val vehicleOutput: VehicleIngredient) : IRecipeWrapper {

    private val slot = guiHelper.createDrawable(VehicleModJei.RECIPE_GUI_VEHICLE, 176, 43, 18, 18)
    private val slotIconColor = guiHelper.createDrawable(VehicleModJei.RECIPE_GUI_VEHICLE, 194, 43, 16, 16)
    private val slotIconNone = guiHelper.createDrawable(VehicleModJei.RECIPE_GUI_VEHICLE, 210, 43, 16, 16)

    override fun getIngredients(ingredients: IIngredients) {
        ingredients.setInputLists(VanillaTypes.ITEM, itemInputs)
        ingredients.setOutput(VanillaTypes.ITEM, vehicleOutput.crate)
    }

    override fun drawInfo(minecraft: Minecraft, recipeWidth: Int, recipeHeight: Int, mouseX: Int, mouseY: Int) {
        val vehicle = vehicleOutput.vehicle

        if (vehicle.canBeColored()) {
            slotIconColor.draw(minecraft, 1, 65)
        } else {
            slotIconNone.draw(minecraft, 1, 65)
        }
        if (vehicle !is EntityPoweredVehicle || vehicle.engineType == EngineType.NONE) {
            slotIconNone.draw(minecraft, 21, 65)
        }
        if (vehicle !is EntityPoweredVehicle || !vehicle.canChangeWheels()) {
            slotIconNone.draw(minecraft, 41, 65)
        }

        for (i in 0 until MATERIAL_SLOTS) {
            slot.draw(minecraft, i * 18, 95)
        }

        minecraft.fontRenderer.drawString(I18n.format("cvmjei.category.vehicle.materials"), 0, 85, 4210752)

        GlStateManager.enableDepth()
        GlStateManager.color(1f, 1f, 1f, 1f)
        GlStateManager.translate(72f, 50f, 100f)
        GlStateManager.rotate(-5f, 1f, 0f, 0f)
        GlStateManager.rotate(Minecraft.getMinecraft().player.ticksExisted + minecraft.renderPartialTicks, 0f, 1f, 0f)
        GlStateManager.scale(SCALE, -SCALE, SCALE)

        val position = GuiWorkstation.DISPLAY_PROPERTIES[vehicleOutput.vehicle.javaClass]
        if (position != null) {
            //Apply vehicle rotations, translations, and scale
            GlStateManager.scale(position.scale, position.scale, position.scale)
            GlStateManager.rotate(position.rotX.toFloat(), 1f, 0f, 0f)
            GlStateManager.rotate(position.rotY.toFloat(), 0f, 1f, 0f)
            GlStateManager.rotate(position.rotZ.toFloat(), 0f, 0f, 1f)
            GlStateManager.translate(position.x, position.y, position.z)
        }
        minecraft.renderManager.getEntityClassRenderObject<EntityVehicle>(vehicleOutput.vehicle.javaClass).doRender(vehicle, 0.0, 0.0, 0.0, 0f, 0f)
        GlStateManager.disableDepth()
    }

    override fun getTooltipStrings(mouseX: Int, mouseY: Int): MutableList<String> {
        val tooltip = ArrayList<String>()

        addSlotTooltip(tooltip, 0, 64, mouseX, mouseY, SlotType.COLOR)
        addSlotTooltip(tooltip, 20, 64, mouseX, mouseY, SlotType.ENGINE)
        addSlotTooltip(tooltip, 40, 64, mouseX, mouseY, SlotType.WHEELS)

        return tooltip
    }

    private fun addSlotTooltip(tooltip: MutableList<String>, slotX: Int, slotY: Int, mouseX: Int, mouseY: Int, type: SlotType) {
        if (mouseX in slotX until slotX + 18 && mouseY in slotY until slotY + 18) {
            if (!type.applicable.apply(vehicleOutput.vehicle)) {
                tooltip.add("${TextFormatting.WHITE}${I18n.format("vehicle.tooltip.${type.registryName}")}")
                tooltip.add("${TextFormatting.GRAY}${I18n.format("vehicle.tooltip.not_applicable")}")
            } else if (type == SlotType.COLOR) {
                tooltip.add("${TextFormatting.AQUA}${I18n.format("vehicle.tooltip.optional")}")
                tooltip.add("${TextFormatting.GRAY}${I18n.format("vehicle.tooltip.${type.registryName}")}")
            }
        }
    }

    companion object {
        const val SCALE = 20f
        const val MATERIAL_SLOTS = 8
    }

    enum class SlotType(val registryName: String, val applicable: Function<EntityVehicle, Boolean>) {
        COLOR("paint_color", Function { it.canBeColored() }),
        ENGINE("engine", Function { it is EntityPoweredVehicle && it.engineType != EngineType.NONE }),
        WHEELS("wheels", Function { it is EntityPoweredVehicle && it.canChangeWheels() })
    }
}